### ISSUE TITLE
Email format not properly validated

### ISSUE THREAD

**Issue by staab** [Julian/jsonschema#403]
```
from jsonschema import Draft4Validator, FormatChecker

validator = Draft4Validator({'type': 'string', 'format': 'email'}, format_checker=FormatChecker())
validator.validate('test@example.com ')
```

The above code doesn't raise an error, but should, at least as far as I understand [RFC 5322](https://tools.ietf.org/html/rfc5322#section-3.4.1).

----------------

Upon further investigation, your [email checking function](https://github.com/Julian/jsonschema/blob/118e6a2412afcb64544c607425fec2ef346880bf/jsonschema/_format.py#L151) is _terrible_. I've been trusting this library to handle the hard parts of validation for me (particularly emails!), but even a _regex_ would have been a better way to validate emails. Just for instance, this passes validation:

```
validator.validate('#@%^%#$@#$@#')
```

I'd really like to see the email validator pass [this suite of tests](https://blogs.msdn.microsoft.com/testing123/2009/02/06/email-address-test-cases/). 

It's fair that you haven't implemented the validation, since the jsonschema spec doesn't require format validation, but partway implementing it, and not warning users (either through raising errors on schemas relying on incompletely-implemented formats, or through documentation), that your format checker doesn't actually validate against the RFC, is pretty bad. My trust in this library has been shaken, and I'll be auditing my code to add custom validation logic where necessary, or switching away entirely.

**Comment by Julian** [comment]
The email validator is intentionally unimplemented, because validating email addresses isn't really a reasonable concept. It's the one and only way that `jsonschema` intentionally deviates from the specification, its done to prevent downstream users from footgunning themselves with a concept that I wish didn't exist in the world, and it was my intention to voice upstream that I think we should remove the email validator. I could have sworn that was noted in the docs, but it's certainly noted in previous issue tickets about this. Patches are certainly welcome.

That being said, your attitude and commentary isn't welcome. Use the library if it helps you, don't if it doesn't, I don't really care about your "trust" in the library.

**Comment by staab** [comment]
After looking through the [validation](https://python-jsonschema.readthedocs.io/en/latest/validate/) section of the docs again and not seeing anything calling out email validation in particular, I went ahead and created #404.

It still seems worth it to me to remove or warn about the email validator, as well as validators that require dependencies or formats that aren't implemented, even though the library is following jsonschema's specification to let that stuff pass. You'd at least get fewer people creating duplicate issues like this one.

**Comment by Julian** [comment]
@staab how do you feel here after that clarification (thanks again!)

The issue with a warning (beyond the fact that the stdlib crippled its own warning system, which is somewhat fixable) is that it's not really possible to distinguish between someone who's *accidentally* forgot to install a dependency from someone who *actually* doesn't mean to use format for validation.

If you still think there's more here though happy to discuss.

**Comment by handrews** [comment]
@staab 
> It seems like it's the fault of the jsonschema spec itself rather than this library, but "failing fast" is generally considered best-practice in software development.

We are working on improving this situation, and others regarding extensibility and modularity (specifically being able to know whether an implementation handles your extensions and failing appropriately if not)  See json-schema-org/json-schema-spec#561 (basics) and json-schema-org/json-schema-spec#563 (formats).  This approach would retain the default permissive behavior, while providing both schema authors and schema implementors control over what they require / accept.

I would advise against sinking too much time into implementation-specific band-aids on this as we do plan to address it in the next draft.  Better to spend that effort adding support for draft-08 as soon as it's out.

**Comment by staab** [comment]
I haven't read the spec (I read half of draft 3 a number of years ago), so my suggestions here are entirely based on my own instincts. If the philosophy of jsonschema conflicts, with my own, for sure go with jsonschema's.

It sounds like you've heard me loud and clear, so I'm pretty satisfied here. I think the `NullFormatChecker` (or at least a `BaseFormatChecker` with all the well-defined ones) is definitely in the spirit of what I've been saying. Also, thanks for your thoughts on `format`; I didn't realize it was meant to be quite so lenient. At any rate, I've got my implementation sorted out, so thanks for going over this, and I look forward to seeing how jsonschema continues to evolve.

**Comment by Julian** [comment]
So, for a concrete resolution here I'm going to close this one, but @staab I've opened #407 to track the possible change in behavior going forward.

Appreciated!
